File: c_sharp/HandicapProgram.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;
using System.IO;

namespace Athensworks.GolfHandicap.Application
{
    class HandicapProgram
    {

        // Please edit this string to match your local file
        private static string filePath = "C:\\Users\\Sidney\\Documents\\Development\\Athensworks.GolfHandicap\\Athensworks.GolfHandicap.Application\\scores.xml";

        private class Round
        {
            public int Score { get; set; }
            public int Slope { get; set; }
            public double Rating { get; set; }
            public int Par { get; set; }
        }
        static void Main(string[] args)
        {
            // get the list of scores
            List<Round> rounds = Scores2List();
            // compute the handicap 15.1
            double handicap = Get_Handicap(rounds);
            // output the result
            Console.Write("The handicap is: " + handicap.ToString());
            Console.ReadLine();
        }
        private static List<Round> Scores2List()
        {
            // load the XML file
            StreamReader scoresFile = new StreamReader(filePath);
            string scoresXML = scoresFile.ReadToEnd();
            scoresFile.Close();
            XmlDocument doc = new XmlDocument();
            doc.LoadXml(scoresXML);
            // defined the XML node containing a round object
            XmlNodeList idNodes = doc.SelectNodes("scores/score");
            // create a new list of rounds
            List<Round> rounds = new List<Round>();
            // populate the list of rounds
            foreach (XmlNode node in idNodes)
            {
                Round newRound = new Round();
                newRound.Score = Convert.ToInt32(node.ChildNodes[0].InnerText);
                newRound.Slope = Convert.ToInt32(node.ChildNodes[1].InnerText);
                newRound.Rating = Convert.ToDouble(node.ChildNodes[2].InnerText);
                newRound.Par = Convert.ToInt32(node.ChildNodes[3].InnerText);
                rounds.Add(newRound);
            }
            // return the list of rounds
            return rounds;
        }
        private static double Get_Handicap(List<Round> ScoreList)
        {
            List<double> diff_List = new List<double>();
            List<double> Result_diffs = new List<double>();
            for (int i = 0; i < ScoreList.Count; i++) // Loop through List with for
            {
                double diff = 0;
                Round c = ScoreList[i];
                diff = ((double)c.Score - c.Rating) * 113 / (double)c.Slope;
                diff_List.Add(diff);
            }
            Result_diffs = get_correct_diffs(diff_List);
            double sum = 0;
            for (int i = 0; i < Result_diffs.Count; i++) // Loop through List with for
            {
                double c = Result_diffs[i];
                sum += c;
            }
            double result = sum / Result_diffs.Count;
            result =  result * .96;
            result = Math.Truncate(10 * result) / 10;
            return result;
        }
        private static List<double> get_correct_diffs(List<double> org_diffs)
        {
            List<Double> result = new List<double>();
            int count = org_diffs.Count;
            int num_diffs = NumLowest(count);
            org_diffs.Sort();
            for (int i = 0; i < num_diffs; i++)
            {
                result.Add(org_diffs[i]);
            }
            return result;
        }
        private static int NumLowest(int num)
        {
            switch (num)
            {
                case 5:
                    return 1;
                    break;
                case 6:
                    return 1;
                    break;
                case 7:
                    return 2;
                    break;
                case 8:
                    return 2;
                    break;
                case 9:
                    return 3;
                    break;
                case 10:
                    return 3;
                    break;
                case 11:
                    return 4;
                    break;
                case 12:
                    return 4;
                    break;
                case 13:
                    return 5;
                    break;
                case 14:
                    return 5;
                    break;
                case 15:
                    return 6;
                    break;
                case 16:
                    return 6;
                    break;
                case 17:
                    return 7;
                    break;
                case 18:
                    return 8;
                    break;
                case 19:
                    return 9;
                    break;
                case 20:
                    return 10;
                    break;
                default:
                    return 0;
                    break;
            }
        }
    }
}

File: visual_basic/HandicapProgram.vb
Imports System.IO
Imports System.Xml

Module HandicapProgram

    ' Please edit this string to match your local file
    Dim filePath As String = "C:\Users\Sidney\Documents\Development\Athensworks.GolfHandicap\Athensworks.GolfHandicap.Application\scores.xml"

    Class Round
        Private _Score As Integer
        Private _Slope As Integer
        Private _Rating As Double
        Private _Par As Integer
        Public Property Score() As Integer
            Get
                Return _Score
            End Get
            Set(ByVal value As Integer)
                _Score = value
            End Set
        End Property
        Public Property Slope() As Integer
            Get
                Return _Slope
            End Get
            Set(ByVal value As Integer)
                _Slope = value
            End Set
        End Property
        Public Property Rating() As Double
            Get
                Return _Rating
            End Get
            Set(ByVal value As Double)
                _Rating = value
            End Set
        End Property
        Public Property Par() As Integer
            Get
                Return _Par
            End Get
            Set(ByVal value As Integer)
                _Par = value
            End Set
        End Property

    End Class

    Sub Main()
        ' get the list of scores
        Dim rounds As List(Of Round) = Scores2List()
        ' compute the handicap 15.1
        Dim handicap As Double = Get_Handicap(rounds)
        ' output the result
        Console.Write("The handicap is: " & handicap.ToString())
        Console.ReadLine()
    End Sub
    Function Scores2List() As List(Of Round)
        ' load the XML file
        Dim scoresFile As New StreamReader(filePath)
        Dim scoresXML As String = scoresFile.ReadToEnd()
        scoresFile.Close()
        Dim doc As New XmlDocument
        doc.LoadXml(scoresXML)
        ' defined the XML node containing a round object
        Dim idNodes As XmlNodeList = doc.SelectNodes("scores/score")
        ' create a new list of rounds
        Dim rounds As New List(Of Round)
        ' populate the list of rounds
        For Each node In idNodes
            Dim newRound As New Round
            newRound.Score = Convert.ToInt32(node.ChildNodes(0).InnerText)
            newRound.Slope = Convert.ToInt32(node.ChildNodes(1).InnerText)
            newRound.Rating = Convert.ToDouble(node.ChildNodes(2).InnerText)
            newRound.Par = Convert.ToInt32(node.ChildNodes(3).InnerText)
            rounds.Add(newRound)
        Next
        ' return the list of rounds
        Scores2List = rounds
    End Function
    Function Get_Handicap(ScoreList As List(Of Round)) As Double
        Dim diff_list As New List(Of Double)
        Dim Result_diffs As New List(Of Double)
        For i = 0 To ScoreList.Count - 1
            Dim diff As Double = 0
            Dim c As Round = ScoreList(i)
            diff = (c.Score - c.Rating) * 113 / c.Slope
            diff_list.Add(diff)
        Next
        Result_diffs = get_correct_diffs(diff_list)
        Dim sum As Double = 0
        For i = 0 To Result_diffs.Count - 1
            Dim c As Double = Result_diffs(i)
            sum += c
        Next
        Dim result As Double = sum / Result_diffs.Count
        result = result * 0.96
        result = Math.Truncate(10 * result) / 10
        Get_Handicap = result
    End Function
    Function get_correct_diffs(org_difffs As List(Of Double))
        Dim result As New List(Of Double)
        Dim count As Integer = org_difffs.Count
        Dim num_diffs As Integer = NumLowest(count)
        org_difffs.Sort()
        For i = 0 To num_diffs - 1
            result.Add(org_difffs(i))
        Next
        get_correct_diffs = result
    End Function
    Function NumLowest(num As Integer) As Integer
        Select Case num
            Case 5
                NumLowest = 1
            Case 6
                NumLowest = 1
            Case 7
                NumLowest = 2
            Case 8
                NumLowest = 2
            Case 9
                NumLowest = 3
            Case 10
                NumLowest = 3
            Case 11
                NumLowest = 4
            Case 12
                NumLowest = 4
            Case 13
                NumLowest = 5
            Case 14
                NumLowest = 5
            Case 15
                NumLowest = 6
            Case 16
                NumLowest = 6
            Case 17
                NumLowest = 7
            Case 18
                NumLowest = 8
            Case 19
                NumLowest = 9
            Case 20
                NumLowest = 10
            Case Else
                NumLowest = 0
        End Select
    End Function

End Module
